packages/rpc/src/index.ts

```diff
@@ -0,0 +1,358 @@
+import { RPC as ToolkitRPC } from "ckb-js-toolkit";
+import {
+  Alert,
+  BannedAddr,
+  Block,
+  BlockEconomicState,
+  CellWithStatus,
+  ChainInfo,
+  Consensus,
+  DryRunResult,
+  Epoch,
+  Hash,
+  Header,
+  HexNumber,
+  HexString,
+  Indexer,
+  LocalNode,
+  OutPoint,
+  RawTxPool,
+  RemoteNode,
+  SyncState,
+  Transaction,
+  TransactionProof,
+  TransactionWithStatus,
+  TxPoolIds,
+  TxPoolInfo,
+  TxPoolVerbosity,
+} from "@ckb-lumos/base";
+
+export type SerializedBlock = HexString;
+export type SerializedHeader = HexString;
+export type OutputsValidator = "default" | "passthrough";
+
+function asyncSleep(ms = 0) {
+  return new Promise((r) => setTimeout(r, ms));
+}
+
+const handler = {
+  get: (target: any, method: string) => {
+    return async (...params: any) => {
+      const result = await target.rpc[method](...params);
+      if (target.indexer) {
+        await target.waitForBlockSync();
+      }
+      return result;
+    };
+  },
+};
+
+class RpcProxy {
+  private rpc: ToolkitRPC;
+  private indexer: Indexer | undefined;
+  private waitForSyncCheckIntervalSeconds: number;
+  private blockDifference: number;
+
+  constructor(
+    uri: string,
+    indexer?: Indexer,
+    {
+      waitForSyncCheckIntervalSeconds = 1,
+      blockDifference = 3,
+    }: {
+      waitForSyncCheckIntervalSeconds?: number;
+      blockDifference?: number;
+    } = {}
+  ) {
+    this.rpc = new ToolkitRPC(uri);
+    this.indexer = indexer;
+    this.waitForSyncCheckIntervalSeconds = waitForSyncCheckIntervalSeconds;
+    this.blockDifference = blockDifference;
+  }
+
+  getProxy() {
+    return new Proxy(this, handler);
+  }
+
+  async waitForBlockSync() {
+    if (!this.indexer) {
+      return;
+    }
+    const header: Header = await this.rpc.get_tip_header();
+    const blockNumber = BigInt(header.number);
+    while (true) {
+      const tip = await this.indexer.tip();
+      if (tip) {
+        const indexedNumber = BigInt(tip.block_number);
+        if (blockNumber - indexedNumber <= this.blockDifference) {
+          // TODO: do we need to handle forks?
+          break;
+        }
+      }
+      await asyncSleep(this.waitForSyncCheckIntervalSeconds * 1000);
+    }
+  }
+}
+
+export class RPC {
+  private rpcProxy: any;
+
+  /**
+   *
+   * @param uri
+   * @param indexer Will `waitForSync` after rpc call if provided.
```

Comment:
Is there a way to opt out of the waitForSync behavior?

Reply:
How about add a new method?

```typescript
resetIndexer(indexer?: Indexer) {
  this.indexer = indexer
}
```

Or you mean exit from a running `waitForSync` ?

Reply:
I just mean that there should a way to disable the `waitForSync` behavior. For the average crowd we want to keep them on the safe side, but for the dangerous tweakers they should be able to kill this feature for more performance.

Reply:
A `resetIndexer` method added.

https://github.com/nervosnetwork/lumos/pull/148/commits/1b345dea9045555f69503b7e2b185f0f290722bf